kong/pdk/vault.lua

```diff
@@ -679,36 +692,32 @@ local function new(self)
     end
 
     for k, v in pairs(config) do
-      if type(v) == "table" then
+      if k ~= "$refs" and type(v) == "table" then
         config[k] = update(v)
       end
     end
 
-    -- This can potentially grow without ever getting
-    -- reset. This will only happen when a user repeatedly changes
-    -- references without ever restarting kong, which sounds
-    -- kinda unlikely, but should still be monitored.
     local refs = config["$refs"]
     if type(refs) ~= "table" or isempty(refs) then
       return config
     end
 
-    local function update_references(refs, target)
-      for field_name, field_value in pairs(refs) do
-        if is_reference(field_value) then
-          local value, err = get_from_cache(field_value)
-          if not value then
-            self.log.notice("error updating secret reference ", field_value, ": ", err)
-          end
-          target[field_name] = value or ''
-        elseif type(field_value) == "table" then
-          update_references(field_value, target[field_name])
+    for field_name, reference in pairs(refs) do
+      if type(reference) == "string" then -- a string reference
+        config[field_name] = get_from_cache(reference)
+
+      elseif isarray(reference) then -- array or set of references
+        for i = 1, #reference do
+          config[field_name][i] = get_from_cache(reference[i])
+        end
+
+      elseif type(reference) == "table" then -- map of references
+        for key, ref in pairs(reference) do
+          config[field_name][key] = get_from_cache(ref)
         end
       end
     end
 
-    update_references(refs, config)
```

Comment:
Why is the recursive update not needed?  Is`update` itself being called recursively?

Reply:
Update itself has been recursive from day 1. It just didn't handle arrays/sets/maps correctly. The recursion can be seen here:
https://github.com/Kong/kong/blob/bd5e9c156774ba1a526c8753efa08855975704d1/kong/pdk/vault.lua#L597

It is also on there after the refactorings (but that removed the k == "$refs" that should not be recursed, this PR adds it back):
https://github.com/Kong/kong/blob/ce9813a7fdf57fa060a841eea59f9dee4f49a41a/kong/pdk/vault.lua#L683